annealing_helper_functions.py

```diff
@@ -76,3 +77,55 @@ def simulated_annealing(graph, function, initial_X, initial_temp, cool, reanneal
                 T = 1.
             
     return X, history
+
+def prob(a, b):
+    return np.exp(a / b)
+
+# Adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+def parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                       iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are ok
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
+
+    # Initialize stuff
+    nsystems = len(initial_temps)
+    Xs = initial_Xs
+    Ts = initial_temps
+    P = [lambda d: np.exp(d / Ts[c]) for c in range(nsystems)]
+    prev_Es = [function(graph, Xs[i]) for i in range(nsystems)]
+    history = [[] for i in xrange(nsystems)]
+
+    for c in xrange(nsystems):
+        history[c].append(prev_Es[c])
+
+    for step in range(iterr):
+        # Run nbefore steps of simulated annealing
+        X_stars = [swap_function(Xs[i], nswaps) for i in xrange(nsystems)]
+        new_Es = [function(graph, X_stars[i]) for i in xrange(nsystems)]
+        delta_Es = [i - j for i, j in zip(new_Es, prev_Es)]
+        
+        # Do a normal update
+        for c in range(nsystems):
+            if np.random.uniform() < P[c](-delta_Es[c]):
+                history[c].append(new_Es[c])
+                Xs[c] = X_stars[c]
+                prev_Es[c] = new_Es[c]
+
+        # Decide which chains, if any, to exchange
+        if step % nbefore == 0:
+            for c in range(nsystems - 1):
+                # Acceptance probability
+                A = min(1, np.exp(((delta_Es[c] - delta_Es[c+1])/Ts[c]) +
+                                  ((delta_Es[c+1] - delta_Es[c])/Ts[c+1])))
```

Comment:
1. I don't know what k is supposed to be in that article. I didn't see any references to it in other material, so I ignored it.

2. Yes, but I did it on the basis of "that's what Pavlos did". I think the idea is that if you can exchange adjacent systems, then in principle you can exchange any system with any other system at any step, but you don't incur the factorial cost. With three systems I doubt it makes a difference, but it's worth investigating when we start running a lot of systems in parallel.

3. Oops